tokamak.py

```diff
@@ -0,0 +1,1190 @@
+        """
+        result = {}
+        for name, segment in segments.items():
+            segment_with_psival = segment.copy()
+            
+            psi_func = self.getPolynomialGridFunc(segment["nx"],
+                                                  segment["psi_start"],
+                                                  segment["psi_end"],
+                                                  grad_lower=segment.get("grad_start", None),
+                                                  grad_upper=segment.get("grad_end", None))
+                
+            segment_with_psival["psi_vals"] = self.make1dGrid(segment["nx"], psi_func)
+            result[name] = segment_with_psival
+        return result
+
+    def setRegions(self, all_regions, segments):
+        """
+        Fill self.regions with EquilibriumRegion objects, 
+        using specifications in all_regions dictionary.
+        
+        Inputs
+        ------
+
+        all_regions   Dictionary containing specification for each region
+        segments      Dictionary of radial segment definitions
+                        nx            Number of radial cells
+                        psi_vals      1D array of psi values, length 2*nx+1
+        
+        Returns
+        -------
+
+        None. Modifies self.regions
+        """
+
+        # Normalisation of the grid cells number
+        # Set N_norm to the total number of grid cells in y
+        setDefault(self.options, 'N_norm', sum([region["ny"]
+                                                for region in all_regions.values()]))
+        
+        self.regions = OrderedDict()
+        
+        # Loop through all regions. For each one create an EquilibriumRegion
+        for name, region in all_regions.items():
+            eqreg = EquilibriumRegion(self,
+                                      name,
+                                      len(region["segments"]), # The number of radial regions
+                                      self.user_options,
+                                      self.options.push(
+                                          {"nx": [segments[seg_name]["nx"]
+                                                  for seg_name in region["segments"]],
+                                           "ny": region["ny"],
+                                           "kind": region["kind"]}),
+                                      # The following arguments are passed through to PsiContour
+                                      region["points"],  # list of Point2D objects on the line
+                                      self.psi,   # Function to calculate the poloidal flux
+                                      region["psi"])
+
+            # Grids of psi values in each segment
+            eqreg.psi_vals = [segments[name]['psi_vals']
+                              for name in region["segments"]]
+                
+            eqreg.separatrix_radial_index = 1
```

Comment:
I did define this as the segment index of the primary separatrix, but I think there's something that might need a bit more work for nonorthogonal spacing to work right in disconnected-double-null:
* `separatrix_radial_index` is used in `globalXInd()` https://github.com/boutproject/hypnotoad2/blob/504cc46cf54a6e96364eb47c24646fab10b6b2b1/mesh.py#L636-L648
* `globalXInd()` is used to set `global_xind` of contours, which is then used to set weightings for spacing functions
https://github.com/boutproject/hypnotoad2/blob/504cc46cf54a6e96364eb47c24646fab10b6b2b1/equilibrium.py#L1663-L1685
This might work out OK, because usually the separatrices are close together, but ideally the `range_lower` and `range_upper` bits would probably use radial distance from the separatrix of the X-point at the lower or upper, respectively, end of the contour rather than necessarily the primary separatrix.

Don't think this comment needs addressing now, but maybe worth bearing in mind for if nonorthogonal grid spacing looks a bit odd for disconnnected double null.